xwiki-platform-core/xwiki-platform-component/xwiki-platform-component-wiki/src/main/java/org/xwiki/component/wiki/internal/bridge/WikiObjectComponentManagerEventListenerProxy.java

```diff
@@ -0,0 +1,214 @@
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.component.manager.ComponentLookupException;
+import org.xwiki.component.manager.ComponentManager;
+import org.xwiki.component.wiki.WikiComponent;
+import org.xwiki.component.wiki.WikiComponentException;
+import org.xwiki.component.wiki.WikiObjectComponentBuilder;
+import org.xwiki.component.wiki.internal.DefaultWikiObjectComponentManagerEventListener;
+import org.xwiki.component.wiki.internal.WikiComponentManagerEventListenerHelper;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.DocumentReferenceResolver;
+import org.xwiki.model.reference.EntityReferenceSerializer;
+import org.xwiki.model.reference.LocalDocumentReference;
+import org.xwiki.model.reference.ObjectReference;
+import org.xwiki.query.Query;
+import org.xwiki.query.QueryManager;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+import com.xpn.xwiki.objects.BaseObjectReference;
+
+/**
+ * This component allows the {@link DefaultWikiObjectComponentManagerEventListener} to easily register new XObject based
+ * components against the {@link WikiComponentManagerEventListenerHelper}.
+ *
+ * @version $Id$
+ * @since 9.5RC1
+ */
+@Component(roles = WikiObjectComponentManagerEventListenerProxy.class)
+@Singleton
+public class WikiObjectComponentManagerEventListenerProxy
+{
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private ComponentManager componentManager;
+
+    @Inject
+    private WikiComponentManagerEventListenerHelper wikiComponentManagerEventListenerHelper;
+
+    @Inject
+    private QueryManager queryManager;
+
+    @Inject
+    private Provider<XWikiContext> xWikiContextProvider;
+
+    @Inject
+    @Named("local")
+    private EntityReferenceSerializer<String> entityReferenceSerializer;
+
+    @Inject
+    @Named("current")
+    private DocumentReferenceResolver<String> currentDocumentReferenceResolver;
+
+    /**
+     * This method is responsible look through every {@link WikiObjectComponentBuilder} and get their role hints, which
+     * is also the class name of the wiki object they should be working with.
+     */
+    private List<String> collectWikiObjectsList()
+    {
+        List<String> wikiObjectsList = new ArrayList<>();
+
+        try {
+            // Get a list of WikiObjectComponentBuilder
+            List<WikiObjectComponentBuilder> componentBuilders =
+                    this.componentManager.getInstanceList(WikiObjectComponentBuilder.class);
+
+            for (WikiObjectComponentBuilder componentBuilder : componentBuilders) {
+                wikiObjectsList.add(this.entityReferenceSerializer.serialize(componentBuilder.getClassReference()));
```

Comment:
Ok, I’ll move to EntityReferences instead of strings in the wikiObjectsList.

Regarding `WikiObjectComponentBuilder#getEntityReference()`, I can absolutely change the definition of the API.